### PR TITLE
consensus: implement double sign risk reduction adr-051

### PR DESCRIPTION
**Pull request by dongsam**
Implementation spec of Double Signing Risk Reduction [ADR-51](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-051-double-signing-risk-reduction.md) by B-Harvest
- Add `DoubleSignCheckHeight` config variable to ConsensusConfig for "How many blocks looks back to check existence of the node's consensus votes when before joining consensus"
- Add `consensus.double_sign_check_height` to `config.toml` and `tendermint node` as flag for set `DoubleSignCheckHeight`
- Set default `consensus.double_sign_check_height` to `0`  ( it could be adjustable in this PR, disable when 0  )

Refs

- [ADR-51](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-051-double-signing-risk-reduction.md)
- [https://github.com/tendermint/tendermint/issues/4059](https://github.com/tendermint/tendermint/issues/4059)
- [https://github.com/tendermint/tendermint/pull/4262](https://github.com/tendermint/tendermint/pull/4262)

* [x]  Wrote tests
* [x]  Updated CHANGELOG_PENDING.md
* [x]  Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
* [x]  Updated relevant documentation (`docs/`) and code comments
* [x]  Re-reviewed `Files changed` in the Github PR explorer
* [ ]  Applied Appropriate Labels

### CODE REVIEW

File: cmd/tendermint/commands/run_node.go [hunk 1]
```diff
@@ -37,6 +37,9 @@ func AddNodeFlags(cmd *cobra.Command) {
 		"genesis_hash",
 		[]byte{},
 		"Optional SHA-256 hash of the genesis file")
+	cmd.Flags().Int64("consensus.double_sign_check_height", config.Consensus.DoubleSignCheckHeight,
+		"How many blocks looks back to check existence of the node's "+
```
Comment (melekes) [inline cmd/tendermint/commands/run_node.go:41]:
```suggestion
		"How many blocks to look back to check existence of the node's "+
```

File: cmd/tendermint/commands/run_node.go [hunk 2]
```diff
@@ -37,6 +37,9 @@ func AddNodeFlags(cmd *cobra.Command) {
...
+			"consensus votes when before joining consensus (disable when 0)")
```
Comment (melekes) [inline cmd/tendermint/commands/run_node.go:42]:
```suggestion
			"consensus votes before joining consensus")
```

File: config/toml.go [hunk 3]
```diff
@@ -378,6 +378,13 @@ timeout_precommit = "{{ .Consensus.TimeoutPrecommit }}"
 timeout_precommit_delta = "{{ .Consensus.TimeoutPrecommitDelta }}"
 timeout_commit = "{{ .Consensus.TimeoutCommit }}"
 
+# How many blocks looks back to check existence of the node's consensus votes when before joining consensus
```
Comment (melekes) [inline config/toml.go:381]:
```suggestion
# How many blocks to look back to check existence of the node's consensus votes before joining consensus
```

File: config/toml.go [hunk 4]
```diff
@@ -378,6 +378,13 @@ timeout_precommit = "{{ .Consensus.TimeoutPrecommit }}"
...
+# When the risk reduction method is on, restarting a validator node will panic
```
Comment (melekes) [inline config/toml.go:382]:
```suggestion
# When non-zero, the node will panic upon restart
# if the same consensus key was used to sign {double_sign_check_height} last blocks.
```

File: config/toml.go [hunk 5]
```diff
@@ -378,6 +378,13 @@ timeout_precommit = "{{ .Consensus.TimeoutPrecommit }}"
...
+# because the node itself voted on consensus with the same consensus key.
+# So, validators should stop the state machine, wait for some blocks, and then restart the state machine to avoid panic.
+# - 0 disable.
```
Comment (melekes) [inline config/toml.go:385]:
```suggestion
```

File: consensus/state.go [hunk 6]
```diff
@@ -354,6 +355,30 @@ func (cs *State) OnStart() error {
 		return err
 	}
 
+	// Double Signing Risk Reduction Logic
+	// triggered only when validator mode(existing privValidator) and DoubleSignCheckHeight set
```
Comment (melekes) [inline consensus/state.go:359]:
```suggestion
```

File: consensus/state.go [hunk 7]
```diff
@@ -354,6 +355,30 @@ func (cs *State) OnStart() error {
 		return err
 	}
 
+	// Double Signing Risk Reduction Logic
```
Comment (melekes) [inline consensus/state.go:358]:
```suggestion
	// Double Signing Risk Reduction
```

File: consensus/state.go [hunk 8]
```diff
@@ -354,6 +355,30 @@ func (cs *State) OnStart() error {
...
+	// triggered only when validator mode(existing privValidator) and DoubleSignCheckHeight set
+	if !(cs.privValidator == nil) && cs.config.DoubleSignCheckHeight > 0 && cs.Height > 0 {
```
Comment (melekes) [inline consensus/state.go:360]:
```suggestion
	if cs.privValidator != nil && cs.config.DoubleSignCheckHeight > 0 && cs.Height > 0 {
```

File: consensus/state.go [hunk 9]
```diff
@@ -354,6 +355,30 @@ func (cs *State) OnStart() error {
...
+		valAddr, err := cs.privValidator.GetPubKey()
+		if err != nil {
+			return err
+		}
+
+		doubleSignCheckHeight := cs.config.DoubleSignCheckHeight
+		if doubleSignCheckHeight > cs.Height {
```
Comment (melekes) [inline consensus/state.go:367]: shouldn't we compare with `cs.Height-1`?

File: consensus/state.go [hunk 10]
```diff
@@ -354,6 +355,30 @@ func (cs *State) OnStart() error {
...
+			doubleSignCheckHeight = cs.Height
+		}
+		for i := int64(1); doubleSignCheckHeight > i; i++ {
+			lastCommit := cs.blockStore.LoadSeenCommit(cs.Height - i)
+			if lastCommit != nil {
+				for _, s := range lastCommit.Signatures {
+					if s.BlockIDFlag == types.BlockIDFlagCommit && bytes.Equal(s.ValidatorAddress, valAddr.Address()) {
+						cs.Logger.Error("Error checking double sign risk reduction logic", "err", ErrDoubleSignRiskReduction)
```
Comment (melekes) [inline consensus/state.go:375]:
```suggestion
						cs.Logger.Info("Found signature from the same key", "sig", s, "idx", sigIdx, "height", cs.Height-i)
```

File: consensus/state.go [hunk 11]
```diff
@@ -354,6 +355,30 @@ func (cs *State) OnStart() error {
...
+		}
+
+		doubleSignCheckHeight := cs.config.DoubleSignCheckHeight
+		if doubleSignCheckHeight > cs.Height {
+			doubleSignCheckHeight = cs.Height
+		}
+		for i := int64(1); doubleSignCheckHeight > i; i++ {
+			lastCommit := cs.blockStore.LoadSeenCommit(cs.Height - i)
+			if lastCommit != nil {
+				for _, s := range lastCommit.Signatures {
```
Comment (melekes) [inline consensus/state.go:373]:
```suggestion
				for sigIdx, s := range lastCommit.Signatures {
```

File: consensus/state.go [hunk 12]
```diff
@@ -36,6 +36,7 @@ var (
 	ErrInvalidProposalSignature = errors.New("error invalid proposal signature")
 	ErrInvalidProposalPOLRound  = errors.New("error invalid proposal POL round")
 	ErrAddingVote               = errors.New("error adding vote")
+	ErrDoubleSignRiskReduction  = errors.New("double sign detected or restarted in DoubleSignCheckHeight")
```
Comment (melekes) [inline consensus/state.go:39]:
```suggestion
	ErrSignatureFoundInPastBlocks  = errors.New("found signature from the same key")
```

File: consensus/state.go [hunk 13]
```diff
@@ -354,6 +355,30 @@ func (cs *State) OnStart() error {
...
+		if doubleSignCheckHeight > cs.Height {
+			doubleSignCheckHeight = cs.Height
+		}
+		for i := int64(1); doubleSignCheckHeight > i; i++ {
+			lastCommit := cs.blockStore.LoadSeenCommit(cs.Height - i)
+			if lastCommit != nil {
+				for _, s := range lastCommit.Signatures {
+					if s.BlockIDFlag == types.BlockIDFlagCommit && bytes.Equal(s.ValidatorAddress, valAddr.Address()) {
+						cs.Logger.Error("Error checking double sign risk reduction logic", "err", ErrDoubleSignRiskReduction)
+						return ErrDoubleSignRiskReduction
```
Comment (melekes) [inline consensus/state.go:376]: I can't find a place where we panic...

Comment (dongsam) [reply consensus/state.go:376]: when return error on there `func (cs *State) OnStart() error`, panic occurs in [`SwitchToConsensus`](https://github.com/tendermint/tendermint/blob/2ac5a559b4c88385e173b673d1f6250ff7c55263/consensus/reactor.go#L125-L127)

log example
```
I[2020-07-24|08:04:15.056] Started node                                 module=main nodeInfo="{ProtocolVersion:{P2P:9 Block:12 App:1} DefaultNodeID:fe0092406535cb943f5e15e3c6c33718089e1d3a ListenAddr:tcp://0.0.0.0:26656 Network:chain-hsa2Bm Version:0.33.6 Channels:40202122233038606100 Moniker:8CBB29C542D2D308 Other:{TxIndex:on RPCAddress:tcp://0.0.0.0:26657}}"
I[2020-07-24|08:04:15.367] Executed block                               module=state height=37 validTxs=0 invalidTxs=0
I[2020-07-24|08:04:15.367] Committed state                              module=state height=37 txs=0 appHash=0000000000000000
panic: Failed to start consensus state: double sign detected or restarted in DoubleSignCheckHeight

conS:
ConsensusState

conR:
ConsensusReactor

goroutine 45 [running]:
github.com/tendermint/tendermint/consensus.(*Reactor).SwitchToConsensus(0xc00025f880, 0xc, 0x1, 0xc0002a8440, 0x6, 0xc0002a8450, 0xc, 0x25, 0xc00258ebc0, 0x20, ...)
	github.com/tendermint/tendermint/consensus/reactor.go:127 +0x2cb
github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).poolRoutine(0xc000b44700, 0xc0024af800)
	github.com/tendermint/tendermint/blockchain/v0/reactor.go:319 +0x11be
created by github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).OnStart
	github.com/tendermint/tendermint/blockchain/v0/reactor.go:110 +0x89
```

File: consensus/state.go (hunk 9 above)
Comment (dongsam) [reply consensus/state.go:367]: If that's the case, I expect this result

File: config/toml.go [hunk 14]
```diff
@@ -378,6 +378,13 @@ timeout_precommit = "{{ .Consensus.TimeoutPrecommit }}"
 timeout_precommit_delta = "{{ .Consensus.TimeoutPrecommitDelta }}"
 timeout_commit = "{{ .Consensus.TimeoutCommit }}"
 
+# How many blocks to look back to check existence of the node's consensus votes before joining consensus
+# When non-zero, the node will panic upon restart
+# if the same consensus key was used to sign {double_sign_check_height} last blocks.
+# because the node itself voted on consensus with the same consensus key.
```
Comment (melekes) [inline config/toml.go:384]:
```suggestion
```

File: consensus/state.go [hunk 15]
```diff
@@ -354,6 +355,29 @@ func (cs *State) OnStart() error {
...
+		valAddr, err := cs.privValidator.GetPubKey()
+		if err != nil {
+			return err
+		}
+
+		doubleSignCheckHeight := cs.config.DoubleSignCheckHeight
+		if doubleSignCheckHeight > cs.Height {
+			doubleSignCheckHeight = cs.Height
+		}
+		for i := int64(1); doubleSignCheckHeight > i; i++ {
```
Comment (melekes) [inline consensus/state.go:369]:
```suggestion
		for i := int64(1); i <= doubleSignCheckHeight; i++ {
```

this is more standard style

File: consensus/state.go [hunk 16]
```diff
@@ -354,6 +355,29 @@ func (cs *State) OnStart() error {
 		return err
 	}
 
+	// Double Signing Risk Reduction
```
Comment (marbar3778) [inline consensus/state.go:358]: is there any way we could get a testcase for this functionality?

Comment (dongsam) [reply consensus/state.go:358]: Yes, I'm writing it now, 
I'm trying to write an integration test using a docker like `test/p2p/fast_sync`. Is that okay?

File: consensus/state.go [hunk 17]
```diff
@@ -354,6 +355,29 @@ func (cs *State) OnStart() error {
...
+	if cs.privValidator != nil && cs.config.DoubleSignCheckHeight > 0 && cs.Height > 0 {
+		valAddr, err := cs.privValidator.GetPubKey()
```
Comment (melekes) [inline consensus/state.go:360]:
```suggestion
		valAddr := cs.privValidatorPubKey.Address()
```

File: consensus/state.go [hunk 18]
```diff
@@ -354,6 +355,29 @@ func (cs *State) OnStart() error {
 		return err
 	}
 
+	// Double Signing Risk Reduction
+	if cs.privValidator != nil && cs.config.DoubleSignCheckHeight > 0 && cs.Height > 0 {
```
Comment (melekes) [inline consensus/state.go:359]:
```suggestion
	if cs.privValidator != nil && cs.privValidatorPubKey != nil && cs.config.DoubleSignCheckHeight > 0 && cs.Height > 0 {
```

**Comment by marbar3778** [comment]
@zmanian @ebuchman @brapse do you have thoughts on this?

**Comment by erikgrinaker** [comment]
I'm late to the game here, so don't let me block this, but I was asked for an opinion. Personally I feel like this misses too many corner cases, and that we probably don't want to weaken the incentives for operators to manage their keys and nodes properly in the first place. Ideally, a private key should only be used by a single node.

From what I understand, the primary use-case here is validator failover. Alternative solutions might be to e.g. allow multiple private keys per validator (in prioritized order) or use some sort of distributed lock service to make sure only a single validator replica is attempting to vote for a given height. I haven't given this much thought though, just a couple of ideas.

**Comment by tessr** [comment]
> I'm late to the game here, so don't let me block this, but I was asked for an opinion.

Keys are the kind of thing that are worth getting right the first time, so I think it's OK if we take a step back just to double-check this approach. In this case, we probably should have looped in the folks tagged above at the ADR step of the process rather than at this step. But ah well - we're learning and getting better at that. 

Going to tag @tarcieri on this one too, since I bet he might have an opinion on it 😉

**Comment by tarcieri** [comment]
I'm a fan of a belt-and-suspenders approach here. There's been a lot of discussion of various ways to better mitigate double signing on the KMS side, but given slashing risk, it's nice to also improve checks on the validator side too.

The only thing I might additionally suggest here is a potential optional enhancement (perhaps a configurable one), would be a sort of "warm up" phase. Concretely, if you had two validators, one of them could sign only the odd numbered blocks, and another the even numbered ones (much like the similar trick for database sharding by primary key). Or for 3 validators, it'd be if their "validator index" matched (block_height % 3), (block_height % 3) + 1, (block_height % 3) + 2 respectively.

This would give some time for validators to signal to each other that they're online, and also a way of assigning a priority so if two (or more) validators do happen to go active at the same time, the lowest priority ones (perhaps the ones with the highest "validator index") can observe the highest priority one is signing, at that point deliberately stop signing, and the highest priority one can witness it before it begins signing every block.

All of that seems like a potential next step for this ADR/PR, which otherwise seems like a decent start.

**Comment by dlguddus** [comment]
> I'm a fan of a belt-and-suspenders approach here. There's been a lot of discussion of various ways to better mitigate double signing on the KMS side, but given slashing risk, it's nice to also improve checks on the validator side too.
> 
> The only thing I might additionally suggest here is a potential optional enhancement (perhaps a configurable one), would be a sort of "warm up" phase. Concretely, if you had two validators, one of them could sign only the odd numbered blocks, and another the even numbered ones (much like the similar trick for database sharding by primary key). Or for 3 validators, it'd be if their "validator index" matched (block_height % 3), (block_height % 3) + 1, (block_height % 3) + 2 respectively.
> 
> This would give some time for validators to signal to each other that they're online, and also a way of assigning a priority so if two (or more) validators do happen to go active at the same time, the lowest priority ones (perhaps the ones with the highest "validator index") can observe the highest priority one is signing, at that point deliberately stop signing, and the highest priority one can witness it before it begins signing every block.
> 
> All of that seems like a potential next step for this ADR/PR, which otherwise seems like a decent start.

Thanks for the opinion. We also agree that allowing multiple validators should be ultimate goal, but achieving it with very high certainty of non-double-signing for any edge case is quite a challenge for us. Because the result of double-signing is catastrophic, we hope the future solution should provide us 100% logical certainty that double-signing will not occur in any edge case. 

We will continue to participate on the next step discussion!

**Comment by dongsam** [comment]
I think `Test Coverage / test-coverage-part-*` doesn't seem deterministic in ci, is there anything else I need to do before Merge?

File: config/config.go [hunk 19]
```diff
@@ -845,6 +847,7 @@ func DefaultConsensusConfig() *ConsensusConfig {
 		CreateEmptyBlocksInterval:   0 * time.Second,
 		PeerGossipSleepDuration:     100 * time.Millisecond,
 		PeerQueryMaj23SleepDuration: 2000 * time.Millisecond,
+		DoubleSignCheckHeight:       int64(10),
```
Comment (erikgrinaker) [inline config/config.go:850]: I think it may be unintuitive to enable this by default. For example, if I'm just playing around with a testnet then I won't be able to do a full chain restart since all nodes will panic. I'd vote for disabling this by default, and validators who need it can enable it.

Comment (dongsam) [reply config/config.go:850]: Yes, it's an option that can cause change and confusion, so I think we need to discuss more the default config, and there's also a way to make sure that enable is the default, but disabled on the testnet config

Comment (marbar3778) [reply config/config.go:850]: Since most nodes are not validators we should go with default off. This is the same thing we do with `pex`, we turn it on by default, but it is recommended to have it off for validators.

Comment (erikgrinaker) [reply config/config.go]: Now that it's off by default (thanks!), we can remove all of the `--consensus.double_sign_check_height 0` arguments in the test scripts and such.

File: consensus/state.go [hunk 20]
```diff
@@ -366,6 +367,25 @@ func (cs *State) OnStart() error {
 		return err
 	}
 
+	// Double Signing Risk Reduction
```
Comment (erikgrinaker) [inline consensus/state.go:370]: Is it actually sufficient to only do this in `OnStart()`? If a node has e.g. fast sync disabled, or for whatever other reason isn't caught up with the chain, then a different validator can submit votes for heights that this node doesn't know about yet. You should be able to reproduce this simply by starting a new node with a duplicate key and fast sync disabled.

I believe we will have to run this check again before casting our first vote, after consensus has caught up with the chain head, or something similar.

Comment (dongsam) [reply consensus/state.go:370]: I agree, that would be a more secure way.
What do you think about adding check logic again in [`gossipDataRoutine`](https://github.com/tendermint/tendermint/blob/master/consensus/reactor.go#L540)(height and round matched) for that? and I could be set the `DoubleSignCheckHeight` as a flag to zero to make it work only once here

Comment (erikgrinaker) [reply consensus/state.go:370]: I'm afraid I'm not very familiar with the intricacies of the consensus module, but I'd say the check should run right before we try to cast our first vote, probably in `State` somewhere, not `Reactor`.

Comment (dlguddus) [reply consensus/state.go:370]: Hi. Erik. This is Hyung from B-Harvest. We are quite confident that the proposed additional check will satisfy raised concern about covering different route of risk.

Because this work has been staled,  could you elaborate more detail on the codebase about your concerns? It will be helpful to understand your opinion.

We will wait several days before we move on with our proposed approach.

Thank you for your suggestion!

Comment (melekes) [inline consensus/state.go:370]: can we extract this into a function please?

Comment (melekes) [reply consensus/state.go:370]: Yeah, it won't work without fast-sync / state-sync. We can create a separate issue for improving it. I doubt there's a clear point in code where "right before we try to cast our first vote" happens, so it might be difficult to pin point.

Comment (dongsam) [reply consensus/state.go:370]: sure, I'll create a separate issue for improving it after merged this fast-sync version

**Comment by melekes** [comment]
@dongsam can you also rebase the PR agaist the latest master?

**Comment by dongsam** [comment]
> @dongsam can you also rebase the PR agaist the latest master?

@melekes of course, you mean rebase/squash with force-push? or merge upstream?

**Comment by melekes** [comment]
> you mean rebase/squash with force-push? or merge upstream?

the former one

**Comment by melekes** [comment]
Hmm... github still says "This branch is out-of-date with the base branch"
